deployments/extensions.go

```diff
@@ -0,0 +1,405 @@
+	}
+	return nil
+}
+
+func (e *Extension) UnInstall(c *kubernetes.Cluster, ui *ui.UI, options *kubernetes.InstallationOptions) error {
+
+	namespace := e.desc.Namespace
+	if namespace == "" {
+		namespace = defaultNamespace
+	}
+	// based on installation type (script/helm/manifest), proceed with uninstallation of each install step
+	for _, step := range e.desc.Uninstall {
+
+		ns := step.Namespace
+		if ns == "" {
+			ns = namespace
+		}
+		switch step.Type {
+		case "helm":
+			// TODO shoud step have a Name too? Could there be multiple helm charts?
+			err := e.unInstallHelmPackage(ui, e.Name, ns)
+			if err != nil {
+				return errors.Wrap(err, "failed to uninstall helm release "+e.Name)
+			}
+		case "manifest":
+			err := e.unInstallManifest(step.Location, ns)
+			if err != nil {
+				return errors.Wrap(err, "failed to uninstall kubernetes manifest from "+step.Location)
+			}
+		case "script":
+			err := e.executeScript(step.Location)
+			if err != nil {
+				return errors.Wrap(err, "failed to install using "+step.Location)
+			}
+		}
+		// delete namespace if it was specific to step
+		if step.Namespace != "" && step.Namespace != namespace {
+			if err := deleteNamespace(c, ui, step.Namespace); err != nil {
+				return err
+			}
+
+		}
+	}
+	// delete namespace if it was specific to extension
+	if e.desc.Namespace != "" {
+		if err := deleteNamespace(c, ui, e.desc.Namespace); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *Extension) Install(c *kubernetes.Cluster, ui *ui.UI, options *kubernetes.InstallationOptions) error {
+
+	namespace := e.desc.Namespace
+	if namespace == "" {
+		namespace = defaultNamespace
+	}
+	// based on installation type (script/helm/manifest), proceed with execution of each install step
+	for _, step := range e.desc.Install {
+		ns := step.Namespace
+		if ns == "" {
+			ns = namespace
+		}
+
+		switch step.Type {
+		case "helm":
+			err := e.installHelmPackage(e.Name, step.Location, ns, step.Values)
+			if err != nil {
+				return errors.Wrap(err, "failed to install helm package from "+step.Location)
+			}
+		case "manifest":
+			err := e.installManifest(step.Location, ns)
+			if err != nil {
+				return errors.Wrap(err, "failed to install kubernetes manifest from "+step.Location)
+			}
+		case "script":
+			err := e.executeScript(step.Location)
+			if err != nil {
+				return errors.Wrap(err, "failed to uninstall using "+step.Location)
+			}
+		}
+		if step.Namespace != "" && step.Namespace != namespace {
+			err := c.LabelNamespace(
+				step.Namespace,
+				kubernetes.FusemlDeploymentLabelKey,
+				kubernetes.FusemlDeploymentLabelValue)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if e.desc.Namespace != "" {
+		err := c.LabelNamespace(
+			e.desc.Namespace,
+			kubernetes.FusemlDeploymentLabelKey,
+			kubernetes.FusemlDeploymentLabelValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO wait for some pod to exist/run? Extra option in the description file
```

Comment:
why need to wait here?

Reply:
We have that kind of code when deploying some other components, take a look at `core.go` for example. It should be fine with helm chart (because I added `--wait` option), but I'm not sure about manifests.

Reply:
ok, now I see it, thanks.

For manifests it is indeed difficult to check for the state when we do not know what kind of workloads are created. Maybe the easier way is just to wait for pods to be in the `Ready` state?

Reply:
Yeah, but which pods?
We can either introduce it as a part of installation step description - just list of pods names to wait for.
It is kind of ugly (from user's POV), but easy to implement.

Or we'd need to actually understand the manifest - parse it for pod definitions. This seems ugly from coding POV.

Or we could do something in between, e.g. _wait for all pods in given namespace_ ... but that could be problematic because such namespace could potentially contain some unrelated deployments...

Reply:
I think we can start with all pods.

Another option is using kustomize to insert specific labels and wait for the pods with that label. But I am not sure if we want to go that way in this PR.

Something to think about, what if we only perform installation from manifests (considering that manifests can be generated from helm/istioctl/(is there something else?) and then we insert the custom labels on those manifests using kustomize. I think this would give an easy way to control the resources created by the extensions.

Reply:
Labels could be good also for listing the extensions. However we would need them for all extension types, not just manifest...

Reply:
I would rather solve this in separate PR